include/libtorrent/session_handle.hpp

```diff
@@ -496,7 +496,7 @@ namespace libtorrent {
 		// can perform a key space traversal with a single RPC per node by adjusting
 		// the target value for each RPC. It has no effect on the returned sample value.
 		// The result is posted as a ``dht_sample_infohashes_alert``.
-		void dht_sample_infohashes(udp::endpoint const& ep, sha1_hash const& target);
+		void dht_sample_infohashes(sha1_hash const& nid, udp::endpoint const& ep, sha1_hash const& target);
```

Comment:
I don't understand what this is meant to be used for. It's just passed through into the alert. I was under the impression you wanted the alert to either contain the node-ID of the node responding, or the node ID of our local node that received the response. I don't see how this is useful.

If it's important to add, I think it should either be as the last parameter with a default value, or as a separate overload. In order to stay backwards compatible. It should also be mentioned in the documentation.

Reply:
if this is meant to be the node ID of the *target*, how is the caller supposed to know this? The DHT response message will include it, we should just pick it up from there.

Reply:
Well... the endpoint is passed through, so I figured the node id should be passed through as well. Of course the caller has to know the node id in the first place, but why wouldn't he/it?

Reply:
Well... the endpoint is passed through, so I figured the node id should be passed through as well. Of course the caller has to know the node id in the first place, but why wouldn't he/it? When dealing with DHT, you always know the other nodes' ids, that's in fact what you'll always use to identify other nodes.

The list of live nodes contains node id + endpoint. Info hash sample replies contain node id + endpoint. Thus, the caller always known the node id. Or am I missing something?

Reply:
you can know the endpoint of a DHT node without knowing its node ID. But more importantly, you don't *need* to know its node ID to send a message to it. Its node ID will be returned in its response. Your comment suggest that *that's* the ID that will come back in the alert, but in your patch that's not the case. The node ID you get back is what you pass in, it doesn't matter what you pass in, because it isn't used for anything.

Reply:
The nodes you send this message to are not necessarily in your routing table. In fact, mostly they probably won't be, you will have learned about them in the response of a previous `dht_sample_infohashes_alert`

Reply:
here's an example use, how I envision it: https://github.com/arvidn/libtorrent/pull/4518

Reply:
Let's look at it from a different angle: to determine if you got a reply to a previously sent request for info hash samples, you need a reference. A node can change its node id, right? So in this case it would actually be beneficial to have the node id unchanged from request to reply. 

> The nodes you send this message to are not necessarily in your routing table. In fact, mostly they probably won't be, you will have learned about them in the response of a previous `dht_sample_infohashes_alert`

That's my point - the node list of a dht_sample_infohashes_alert contains both node id and endpoint.

Reply:
> to determine if you got a reply to a previously sent request for info hash samples, you need a reference to match them

This is typically handled at a lower layer, with the transaction ID. If you need some kind of "user data" to tie a response to the request, I don't think using a `sha1_hash` is appropriate. Especially not suggesting that it somehow reflects the node ID the response was received from (since it doesn't, right now).

> That's my point - the node list of a dht_sample_infohashes_alert contains both node id and endpoint for each node. So you know the node id.

I don't think that's a good reason to require it to be passed in. It's not used for anything, so why require the user to pass it in? If you need a "user data" cookie, it should be optional.

wouldn't the endpoint work just as well as such cookie?